File: dns_orchestrator.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net"
	"net/http"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"sync"
	"time"
)

// ------------------------------
// Paths and constants
// ------------------------------
var (
	baseDir      = "."
	configDir    = filepath.Join(baseDir, "config")
	runtimeDir   = filepath.Join(baseDir, "runtime")
	resultDir    = filepath.Join(baseDir, "result")
	dnsDBFile    = filepath.Join(resultDir, "dns.json")
	activePool   = filepath.Join(runtimeDir, "active.json")
	clientDBFile = filepath.Join(runtimeDir, "clients.json")
	domainsFile  = filepath.Join(configDir, "domains.conf")
	clientsFile  = filepath.Join(configDir, "clients.conf")
	geoFile      = filepath.Join(configDir, "geo.conf")

	maxLatency     = 800
	rotateInterval = 10 * time.Second
	apiPort        = 9180
	webPort        = 9181
	tcpPort        = 9055

	mu sync.Mutex
)

// ------------------------------
// Types
// ------------------------------
type DNS struct {
	IP      string `json:"ip"`
	Latency int    `json:"latency"`
	IPv6    bool   `json:"ipv6"`
	Anycast bool   `json:"anycast"`
	Region  string `json:"region,omitempty"`
}

type Client struct {
	IP   string `json:"ip"`
	Tag  string `json:"tag"`
	Geo  string `json:"geo"`
}

// ------------------------------
// Utility functions
// ------------------------------
func ensureDir(path string) {
	if _, err := os.Stat(path); os.IsNotExist(err) {
		os.MkdirAll(path, 0755)
	}
}

func ensureFile(file, defaultContent string) {
	if _, err := os.Stat(file); os.IsNotExist(err) {
		ioutil.WriteFile(file, []byte(defaultContent), 0644)
	}
}

func loadJSON(file string, v interface{}) {
	data, err := ioutil.ReadFile(file)
	if err == nil {
		json.Unmarshal(data, v)
	}
}

func saveJSON(file string, v interface{}) {
	data, _ := json.MarshalIndent(v, "", "  ")
	ioutil.WriteFile(file, data, 0644)
}

// ------------------------------
// DNS Pool Management
// ------------------------------
func selectActivePool() []DNS {
	db := []DNS{}
	loadJSON(dnsDBFile, &db)
	var pool []DNS
	for _, d := range db {
		if d.Latency <= maxLatency {
			pool = append(pool, d)
		}
	}
	sort.Slice(pool, func(i, j int) bool { return pool[i].Latency < pool[j].Latency })
	return pool
}

func rotatePool(pool []DNS) []DNS {
	if len(pool) <= 1 {
		return pool
	}
	return append([]DNS{pool[len(pool)-1]}, pool[:len(pool)-1]...)
}

func updatePool() {
	mu.Lock()
	defer mu.Unlock()
	pool := selectActivePool()
	pool = rotatePool(pool)
	saveJSON(activePool, pool)
}

// ------------------------------
// Client Detection and Geo
// ------------------------------
func autoDetectClient(ip string) {
	mu.Lock()
	defer mu.Unlock()

	clients := []Client{}
	loadJSON(clientDBFile, &clients)
	for _, c := range clients {
		if c.IP == ip {
			return
		}
	}

	tag := detectClientTag(ip)
	geo := detectGeo(ip)

	c := Client{IP: ip, Tag: tag, Geo: geo}
	clients = append(clients, c)
	saveJSON(clientDBFile, clients)
}

func detectClientTag(ip string) string {
	lines, _ := ioutil.ReadFile(clientsFile)
	for _, line := range strings.Split(string(lines), "\n") {
		if strings.TrimSpace(line) == "" {
			continue
		}
		parts := strings.Fields(line)
		if len(parts) == 2 {
			subnet := parts[0]
			tag := parts[1]
			if subnet == "0.0.0.0/0" {
				return tag
			}
			if ipInSubnet(ip, subnet) {
				return tag
			}
		}
	}
	return "DEFAULT"
}

func detectGeo(ip string) string {
	lines, _ := ioutil.ReadFile(geoFile)
	for _, line := range strings.Split(string(lines), "\n") {
		if strings.TrimSpace(line) == "" {
			continue
		}
		parts := strings.Fields(line)
		if len(parts) == 2 {
			region := parts[0]
			subnet := parts[1]
			if subnet == "0.0.0.0/0" {
				return region
			}
			if ipInSubnet(ip, subnet) {
				return region
			}
		}
	}
	return "UNKNOWN"
}

// ------------------------------
// IP/Subnet utils
// ------------------------------
func ipInSubnet(ipStr, subnetStr string) bool {
	ip := net.ParseIP(ipStr)
	_, subnet, err := net.ParseCIDR(subnetStr)
	if err != nil || ip == nil {
		return false
	}
	return subnet.Contains(ip)
}

// ------------------------------
// API / Web
// ------------------------------
func apiHandler(w http.ResponseWriter, r *http.Request) {
	mu.Lock()
	defer mu.Unlock()
	pool := []DNS{}
	clients := []Client{}
	loadJSON(activePool, &pool)
	loadJSON(clientDBFile, &clients)
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(map[string]interface{}{
		"dns_pool": pool,
		"clients":  clients,
	})
}

func webHandler(w http.ResponseWriter, r *http.Request) {
	mu.Lock()
	defer mu.Unlock()
	pool := []DNS{}
	clients := []Client{}
	loadJSON(activePool, &pool)
	loadJSON(clientDBFile, &clients)
	w.Header().Set("Content-Type", "text/html; charset=utf-8")
	fmt.Fprintf(w, "<html><body><h2>DNS Orchestrator</h2><h3>Active DNS</h3><pre>%v</pre><h3>Clients</h3><pre>%v</pre></body></html>", pool, clients)
}

// ------------------------------
// TCP Client Watcher
// ------------------------------
func tcpWatcher() {
	ln, err := net.Listen("tcp", fmt.Sprintf(":%d", tcpPort))
	if err != nil {
		panic(err)
	}
	for {
		conn, _ := ln.Accept()
		ip, _, _ := net.SplitHostPort(conn.RemoteAddr().String())
		autoDetectClient(ip)
		conn.Close()
	}
}

// ------------------------------
// Main
// ------------------------------
func main() {
	// Initialize directories and files
	ensureDir(configDir)
	ensureDir(runtimeDir)
	ensureDir(resultDir)

	ensureFile(dnsDBFile, `[
  {"ip":"1.1.1.1","latency":40,"ipv6":false,"anycast":true,"region":"DEFAULT"},
  {"ip":"8.8.8.8","latency":55,"ipv6":false,"anycast":true,"region":"DEFAULT"},
  {"ip":"2606:4700:4700::1111","latency":45,"ipv6":true,"anycast":true,"region":"DEFAULT"}
]`)
	ensureFile(domainsFile, "example.com\ngoogle.com\ncloudflare.com\n")
	ensureFile(clientsFile, "0.0.0.0/0 DEFAULT\n")
	ensureFile(geoFile, "DEFAULT 0.0.0.0/0\n")
	ensureFile(clientDBFile, "[]")
	ensureFile(activePool, "[]")

	// Update pool periodically
	go func() {
		for {
			updatePool()
			time.Sleep(rotateInterval)
		}
	}()

	// Start API and Web servers
	http.HandleFunc("/api", apiHandler)
	go http.ListenAndServe(fmt.Sprintf(":%d", apiPort), nil)

	http.HandleFunc("/", webHandler)
	go http.ListenAndServe(fmt.Sprintf(":%d", webPort), nil)

	// TCP Client Watcher
	go tcpWatcher()

	fmt.Println("Full DNS Orchestrator Go started")
	fmt.Printf("API: %d, Web: %d, TCP client detection: %d\n", apiPort, webPort, tcpPort)

	select {} // Block forever
}

File: dns_orchestrator.rs
use std::fs::{self, File};
use std::io::{Read, Write};
use std::net::{TcpListener, TcpStream};
use std::sync::{Arc, Mutex};
use std::thread;
use std::time::Duration;
use warp::Filter;
use serde::{Serialize, Deserialize};

#[derive(Serialize, Deserialize, Clone)]
struct DNS {
    ip: String,
    latency: u32,
    ipv6: bool,
    anycast: bool,
    region: String,
    domains: Vec<String>,
}

#[derive(Serialize, Deserialize, Clone)]
struct Client {
    ip: String,
    tag: String,
    geo: String,
    domains: Vec<String>,
}

const BASE_DIR: &str = ".";
const CONFIG_DIR: &str = "./config";
const RUNTIME_DIR: &str = "./runtime";
const RESULT_DIR: &str = "./result";

const DNS_DB_FILE: &str = "./result/dns.json";
const ACTIVE_POOL_FILE: &str = "./runtime/active.json";
const CLIENT_DB_FILE: &str = "./runtime/clients.json";
const CLIENTS_FILE: &str = "./config/clients.conf";
const GEO_FILE: &str = "./config/geo.conf";
const DOMAINS_FILE: &str = "./config/domains.conf";

const MAX_LATENCY: u32 = 800;
const ROTATE_INTERVAL: u64 = 10;
const TCP_PORT: u16 = 9055;
const API_PORT: u16 = 9180;
const WEB_PORT: u16 = 9181;

// ------------------------------
// Utility functions
// ------------------------------
fn ensure_dir(path: &str) {
    if fs::metadata(path).is_err() {
        fs::create_dir_all(path).unwrap();
    }
}

fn ensure_file(path: &str, default: &str) {
    if fs::metadata(path).is_err() {
        let mut f = File::create(path).unwrap();
        f.write_all(default.as_bytes()).unwrap();
    }
}

fn load_json<T: for<'de> Deserialize<'de>>(path: &str) -> Vec<T> {
    let mut data = String::new();
    File::open(path).unwrap().read_to_string(&mut data).unwrap();
    serde_json::from_str(&data).unwrap_or(Vec::new())
}

fn save_json<T: Serialize>(path: &str, v: &Vec<T>) {
    let data = serde_json::to_string_pretty(v).unwrap();
    let mut f = File::create(path).unwrap();
    f.write_all(data.as_bytes()).unwrap();
}

// ------------------------------
// DNS Pool Management
// ------------------------------
fn select_active_pool(db: &Vec<DNS>) -> Vec<DNS> {
    let mut pool: Vec<DNS> = db.iter()
        .cloned()
        .filter(|d| d.latency <= MAX_LATENCY)
        .collect();
    pool.sort_by_key(|d| d.latency);
    pool
}

fn rotate_pool(pool: &mut Vec<DNS>) {
    if pool.len() > 1 {
        let last = pool.pop().unwrap();
        pool.insert(0, last);
    }
}

// ------------------------------
// Client Detection & Geo Routing
// ------------------------------
fn auto_detect_client(ip: String, clients: &Arc<Mutex<Vec<Client>>>, domains: &Vec<String>) {
    let mut clients_lock = clients.lock().unwrap();
    if !clients_lock.iter().any(|c| c.ip == ip) {
        let tag = detect_client_tag(&ip);
        let geo = detect_geo(&ip);
        clients_lock.push(Client { ip: ip.clone(), tag, geo, domains: domains.clone() });
        save_json(CLIENT_DB_FILE, &*clients_lock);
    }
}

fn detect_client_tag(_ip: &str) -> String {
    // simple static for now
    "DEFAULT".to_string()
}

fn detect_geo(_ip: &str) -> String {
    // simple static for now
    "DEFAULT".to_string()
}

// ------------------------------
// TCP Client Watcher
// ------------------------------
fn tcp_watcher(clients: Arc<Mutex<Vec<Client>>>, domains: Vec<String>) {
    let listener = TcpListener::bind(("0.0.0.0", TCP_PORT)).unwrap();
    for stream in listener.incoming() {
        if let Ok(_s) = stream {
            let ip = _s.peer_addr().unwrap().ip().to_string();
            auto_detect_client(ip, &clients, &domains);
        }
    }
}

// ------------------------------
// Web & API
// ------------------------------
async fn start_webapi(active_pool: Arc<Mutex<Vec<DNS>>>, clients: Arc<Mutex<Vec<Client>>>) {
    let pool_filter = warp::any().map(move || active_pool.clone());
    let clients_filter = warp::any().map(move || clients.clone());

    let api = warp::path("api").map(move || {
        let pool = active_pool.lock().unwrap();
        let cls = clients.lock().unwrap();
        warp::reply::json(&serde_json::json!({
            "dns_pool": *pool,
            "clients": *cls
        }))
    });

    let web = warp::path::end().map(move || {
        let pool = active_pool.lock().unwrap();
        let cls = clients.lock().unwrap();
        warp::reply::html(format!(
            "<html><body><h2>DNS Orchestrator</h2><h3>Active DNS</h3><pre>{:?}</pre><h3>Clients</h3><pre>{:?}</pre></body></html>",
            *pool, *cls
        ))
    });

    let routes = api.or(web);
    warp::serve(routes).run(([0,0,0,0], WEB_PORT)).await;
}

// ------------------------------
// Main
// ------------------------------
#[tokio::main]
async fn main() {
    // Directories and files
    ensure_dir(CONFIG_DIR);
    ensure_dir(RUNTIME_DIR);
    ensure_dir(RESULT_DIR);

    ensure_file(DNS_DB_FILE, r#"[
        {"ip":"1.1.1.1","latency":40,"ipv6":false,"anycast":true,"region":"DEFAULT","domains":["example.com"]},
        {"ip":"8.8.8.8","latency":55,"ipv6":false,"anycast":true,"region":"DEFAULT","domains":["google.com"]},
        {"ip":"2606:4700:4700::1111","latency":45,"ipv6":true,"anycast":true,"region":"DEFAULT","domains":["cloudflare.com"]}
    ]"#);

    ensure_file(CLIENT_DB_FILE, "[]");
    ensure_file(DOMAINS_FILE, "example.com\ngoogle.com\ncloudflare.com\n");

    let domains_data = fs::read_to_string(DOMAINS_FILE).unwrap();
    let domains: Vec<String> = domains_data.lines().map(|s| s.to_string()).collect();

    let clients = Arc::new(Mutex::new(Vec::<Client>::new()));
    let active_pool = Arc::new(Mutex::new(Vec::<DNS>::new()));

    // TCP watcher thread
    let clients_clone = Arc::clone(&clients);
    let domains_clone = domains.clone();
    thread::spawn(move || tcp_watcher(clients_clone, domains_clone));

    // Pool updater thread
    let active_clone = Arc::clone(&active_pool);
    thread::spawn(move || {
        loop {
            let db = load_json::<DNS>(DNS_DB_FILE);
            let mut pool = select_active_pool(&db);
            rotate_pool(&mut pool);
            save_json(ACTIVE_POOL_FILE, &pool);
            let mut lock = active_clone.lock().unwrap();
            *lock = pool;
            thread::sleep(Duration::from_secs(ROTATE_INTERVAL));
        }
    });

    // Start web + API
    start_webapi(active_pool, clients).await;
}